File: sdk/npm/src/providers/tempgo-email.ts
import { InternalEmailInfo, Email, Channel } from "../types";
import { normalizeEmail } from "../normalize";
import { fetchWithTimeout } from "../retry";

const CHANNEL: Channel = "tempgo-email";
const API_BASE = "https://tempgo.email";

interface CreateResponse {
  email?: string;
  expires_at?: string;
  expires_in?: number;
  mailbox_id?: string;
  token?: string;
}

interface InboxResponse {
  email?: string;
  expires_in?: number;
  messages?: Record<string, unknown>[];
}

/**
 * 创建 tempgo.email 临时邮箱
 * API: POST /api/generate（无 body，不发送 Content-Type header）
 * token 存储为 mailbox_id
 */
export async function generateEmail(): Promise<InternalEmailInfo> {
  const response = await fetchWithTimeout(`${API_BASE}/api/generate`, {
    method: "POST",
    headers: { Accept: "application/json" },
  });
  if (!response.ok) {
    throw new Error(`tempgo-email: 创建邮箱失败 ${response.status}`);
  }

  const resp = (await response.json()) as CreateResponse;
  const address = String(resp.email || "").trim();
  const mailboxId = String(resp.mailbox_id || "").trim();

  if (!address || !address.includes("@")) {
    throw new Error("tempgo-email: 创建邮箱响应缺少有效地址");
  }
  if (!mailboxId) {
    throw new Error("tempgo-email: 创建邮箱响应缺少 mailbox_id");
  }

  return {
    channel: CHANNEL,
    email: address,
    token: mailboxId,
  };
}

/**
 * 获取 tempgo.email 邮件列表
 * API: GET /api/inbox?email={email}&mailbox_id={mailbox_id}
 * @param token - mailbox_id（创建邮箱时返回）
 * @param email - 邮箱地址
 */
export async function getEmails(
  token: string,
  email: string,
): Promise<Email[]> {
  if (!token) {
    throw new Error("tempgo-email: mailbox_id（token）不能为空");
  }

  const address = String(email || "").trim();
  if (!address) {
    throw new Error("tempgo-email: 邮箱地址为空");
  }

  const params = new URLSearchParams({ email: address, mailbox_id: token });
  const response = await fetchWithTimeout(
    `${API_BASE}/api/inbox?${params.toString()}`,
    {
      method: "GET",
      headers: { Accept: "application/json" },
    },
  );
  if (!response.ok) {
    throw new Error(`tempgo-email: 获取邮件失败 ${response.status}`);
  }

  const resp = (await response.json()) as InboxResponse;
  const messages = Array.isArray(resp.messages) ? resp.messages : [];

  return messages.map((msg) => {
    const raw: Record<string, unknown> = {
      id: msg.id || "",
      from: msg.from || msg.sender || msg.from_address || "",
      to: msg.to || address,
      subject: msg.subject || "",
      text: msg.text || msg.body || "",
      html: msg.html || msg.body_html || "",
      date: msg.date || msg.received_at || msg.created_at || "",
      isRead: msg.is_read || false,
    };
    return normalizeEmail(raw, address);
  });
}

File: sdk/rust/src/providers/tempgo_email.rs
/*!
 * tempgo-email — https://tempgo.email
 * 流程：POST /api/generate 创建邮箱（无 body，不发送 Content-Type）
 * 收信：GET /api/inbox?email={email}&mailbox_id={token}
 */

use crate::config::{block_on, http_client};
use crate::normalize::normalize_email;
use crate::types::{Channel, Email, EmailInfo};
use serde_json::{json, Value};

const API_BASE: &str = "https://tempgo.email";

/// 创建 tempgo.email 临时邮箱
pub fn generate_email() -> Result<EmailInfo, String> {
    block_on(async {
        let client = http_client();

        let resp = client
            .post(format!("{}/api/generate", API_BASE))
            .send()
            .await
            .map_err(|e| format!("tempgo-email: 创建邮箱失败: {}", e))?;

        if !resp.status().is_success() {
            return Err(format!(
                "tempgo-email: 创建邮箱失败 HTTP {}",
                resp.status()
            ));
        }

        let body: Value = resp
            .json()
            .await
            .map_err(|e| format!("tempgo-email: 解析响应失败: {}", e))?;

        let email = body["email"]
            .as_str()
            .ok_or("tempgo-email: 响应中未找到邮箱地址")?
            .to_string();

        let mailbox_id = body["mailbox_id"]
            .as_str()
            .ok_or("tempgo-email: 响应中未找到 mailbox_id")?
            .to_string();

        Ok(EmailInfo {
            channel: Channel::TempgoEmail,
            email,
            token: Some(mailbox_id),
            expires_at: None,
            created_at: None,
        })
    })
}

/// 获取 tempgo.email 邮件列表
pub fn get_emails(token: &str, email: &str) -> Result<Vec<Email>, String> {
    let address = email.trim();
    if address.is_empty() {
        return Err("tempgo-email: 邮箱地址为空".into());
    }

    block_on(async {
        let client = http_client();

        let resp = client
            .get(format!(
                "{}/api/inbox?email={}&mailbox_id={}",
                API_BASE, address, token
            ))
            .send()
            .await
            .map_err(|e| format!("tempgo-email: 获取邮件失败: {}", e))?;

        if !resp.status().is_success() {
            return Err(format!(
                "tempgo-email: 获取邮件失败 HTTP {}",
                resp.status()
            ));
        }

        let body: Value = resp
            .json()
            .await
            .map_err(|e| format!("tempgo-email: 解析响应失败: {}", e))?;

        let messages = match body["messages"].as_array() {
            Some(arr) => arr.clone(),
            None => return Ok(Vec::new()),
        };

        if messages.is_empty() {
            return Ok(Vec::new());
        }

        let mut out = Vec::with_capacity(messages.len());
        for msg in &messages {
            let raw = json!({
                "id": msg.get("id"),
                "from": msg.get("from").or_else(|| msg.get("sender")),
                "to": address,
                "subject": msg.get("subject"),
                "date": msg.get("date").or_else(|| msg.get("received_at")),
                "html": msg.get("html").or_else(|| msg.get("body_html")).and_then(|v| v.as_str()).unwrap_or(""),
                "text": msg.get("text").or_else(|| msg.get("body_plain")).and_then(|v| v.as_str()).unwrap_or(""),
            });
            out.push(normalize_email(&raw, address));
        }
        Ok(out)
    })
}
